### DISCUSSION TITLE
How to serve a static directory?

### DISCUSSION THREAD

**Discussion by flappyBug** [loco-rs/loco#129] [Q&A] [answered]
Because there's no opinion on the frontend part, in my use case, I'd like to serve frontend as static files. In axum, this can be achieve by 
```rust
    Router::new()
        .nest_service("/assets", get_service(ServeDir::new("dist/assets")))
        .fallback_service(get_service(ServeFile::new("dist/index.html")))
```

Is this possible in loco? Thanks

**Comment by kaplanelad** [comment, marked as answer]
Hey @flappyBug 

you can implement the [after_routes](https://github.com/loco-rs/loco/blob/master/src/app.rs#L120) hook. 
The hook provides access to the Axum router, making both the `nest_service` and `fallback_service` available for your use.

For example:

```rust

pub struct App;
#[async_trait]
impl Hooks for App {
 ....

fn after_routes(router: AxumRouter, _ctx: &AppContext) -> Result<AxumRouter> {
        let router = router.nest_service("/assets", get_service(ServeDir::new("dist/assets")))
        .fallback_service(get_service(ServeFile::new("dist/index.html")));
        Ok(router)
    }
....
}

```

**Comment by kaplanelad** [comment]
@flappyBug 

After a brief discussion with @jondot, we've come to the realization that supporting static asset serving can be seamlessly handled as a middleware in Loco, without the need to rely on the hook in this particular scenario.

for example, 
we can add in the middleware config section: 
```yaml
middlewares:
   static:
       folder: frontent/dist
       fallback: frontent/dist/index.html 
...
```

and loco will will implement the `nest_service` and `fallback_service`. 

what do you think?

**Comment by kaplanelad** [comment]
Thanks @flappyBug for bringing up this question. We believe it's essential to include this in the Loco library. I've added a static middleware to address this scenario. Check out the details in the https://github.com/loco-rs/loco/pull/134.

**Comment by flappyBug** [comment]
@kaplanelad Thank you for adopting this feature. It's fantastic to see official support for static asset serving in loco. Regarding the configuration entry, introducing a prefix might be beneficial (alternatively, placing assets in a subfolder could serve as a workaround). One aspect I'm uncertain about with axum is the feasibility of having exclusive endpoints for the fallback service. Specifically, I'm interested in routing all fallbacks to index.html, with the exception of paths under `/api/*`. This way, our backend can manage 404 errors for `/api/*` paths, while the frontend handles all others.
